volatility/framework/plugins/mac/lsmod.py

```diff
@@ -38,12 +38,29 @@ def list_modules(cls, context: interfaces.context.ContextInterface, layer_name:
         """
         kernel = contexts.Module(context, darwin_symbols, layer_name, 0)
 
+        kernel_layer = context.layers[layer_name]
+
         kmod_ptr = kernel.object_from_symbol(symbol_name = "kmod")
 
-        # TODO - use smear-proof list walking API after dev release
         kmod = kmod_ptr.dereference().cast("kmod_info")
-        while kmod != 0:
+
+        yield kmod
+
+        kmod = kmod.next
+
+        seen = set()
+
+        while kmod != 0 and \
+              kmod not in seen and \
+              len(seen) < 1024:
+
+            if not kernel_layer.is_valid(kmod.dereference().vol.offset, kmod.dereference().vol.size):
```

Comment:
I think that dereferencing the pointer and takes its offset would be the same as just `kmod`?  It may make more sense so that it matches the size, so not a big deal

Reply:
I had issues with this. I believe we talked about this in the past, but then I don't remember the resolution. Basically, when you have a pointer, if you do variable.vol.{offset, size} you get the values for the actual pointer, meaning the value of the address its pointing to and then 4 or 8 for the size depending on 32 or 64 bit layer. When you do .dereference() you get the actual address and size of whatever object the pointer is referencing.

Reply:
So, you're entirely right, I think I wasn't clear in what I was say:
```
kmod (pointer value)
kmod.vol.offset (offset of pointer)
kmod.dereference() (object at pointer value)
kmod.dereference().vol.offset (offset of object at pointer value = pointer value)
```
So essentially `kmod.vol.offset` is the wrong value to use, but `kmod` should be identical to `kmod.dereference().vol.offset` only without having to construct the dereferenced object to get it.  Since you shortly after use `kmod.deferefence().vol.size` which we can't use that trick for, the new object will get created regardless.  I might also suggest putting `kmod.deference()` into a variable, so we don't have to construct the object twice.  It's a small optimization, but if this is run many times over it might add up?  As I say, not a blocker and absolutely happy for you to merge it as is, but figured I'd mention it so that we're aware of the trade-offs...  5:)